### DISCUSSION TITLE
何を開発するか・何の技術スタックを使って開発するか・誰が開発するか etc...

### DISCUSSION THREAD

**Discussion by tsmiyamoto** [YAKIONIGIRIs/Mapda3#1] [Ideas]
- 何を開発するか
- 何の技術スタックを使って開発するか
- 誰が開発するか

などについて話し合うスレ

**Comment by tsmiyamoto** [comment]
マップの位置情報と写真を1つのデータとしてブロックチェーンに組み込む機能

**Reply by tsmiyamoto** [reply]
ここからここまで走りました、が1つのブロックで、1秒1秒チェーンにしていく必要はない

**Reply by tsmiyamoto** [reply]
「家ばれるじゃん！」に対しては、なんかいい感じに（ランドマークからランドマークみたいな）

**Comment by tsmiyamoto** [comment]
対向車と写真を合成して、あたかもストリートビューみたいな画像を作れる機能

**Reply by tsmiyamoto** [reply]
アプリケーションとして作り込むまでもない（実際の走行データがあるわけではないので）

デモで、「こんな感じになりますよー」と見せられるレベルの
- ブロックチェーン上から緯度経度及びその時の写真を撮ってくるプログラム
- 対向車との画像を繋ぎ合わせるアルゴリズム

ができていればいいんじゃない？

**Comment by tsmiyamoto** [comment]
ドラレコ・ナビ・車が一体となって、これが実現できますよ想定

**Comment by tsmiyamoto** [comment]
切り口
**ドラレコとマップって、今結びついて無くないすか？**
YouTubeでよく上がってるどこからどこまで走りました、みたいな動画も、マップ情報ないよね～

根幹
- 走行経路と走行動画をブロックチェーン上に刻む

そのうえで実現できること
- 同乗者が楽しめるよう、走行中に撮った写真とリンクできる機能
- 「今、いろは坂の紅葉どうなんだ？」みたいな誰かが行ったあの場所をトークンで買える
  - インスタで誰があげたの見ればよくね？とか言われそうやけどまあいいや


走行データは常に刻むが、SNS公開するかは自由
公開して、その走行情報を誰かがトークン使って買うことも可能

今のGoogleマップは「そこに行く最短経路」だけを提示してくれるが、Web3上に上がったマップによって「神庭の滝に行くまでに、美味しいカレーによるルートもあれば、美味しい蕎麦ルートもありますよ」を提案してくれる。そこでどこかのユーザーが選択したらトークンのやり取りが発生。
美味しい蕎麦ありましたよ、みたいな感じでマップに付加価値をつけていくと、トークンが得られやすくなる。

マップ閲覧でもトークン利用するし、最終的に「この経路」ってなればプラストークン

**Comment by tsmiyamoto** [comment]
Thanks coin
対向車の貢献によって、この時間のこの場所のストリートビューが作成できたよ！に対してもトークン発行

**Comment by tsmiyamoto** [comment]
**走行経路と走行動画をブロックチェーン上に刻む**
バックエンドとして、こういうプログラムが動いて収集されていきますよ、ができていればOKだと思う
それが、実際にチェーンに刻まれて行ってますよってことが重要

**Reply by tsmiyamoto** [reply]
バックエンド要らなくない？
要らない、は言い過ぎだけど、実際の走行データ取得できるわけではないので、「走行した経路がこのように反映されます。ユーザーは任意に画像やコメントを追加し、マップに価値をつけることができます」みたいな。

**Reply by tsmiyamoto** [reply]
モックデータとして、みんなの散歩動画と経路情報を合わせたやつを10ぐらい用意
フロントでデモするときに、ランダムに仮で出てくるように

**Reply by tsmiyamoto** [reply]
美観地区の中みたいな、車が走行できない部分は散歩になるから、いっそそっちに対応させる（というのは今後の展望ぐらいにしとこう）

**Reply by tsmiyamoto** [reply]
@BB-Cotton 担当

**Reply by tsmiyamoto** [reply]
~~サンプルコードとして、次のができるようなものを11/5(日)には完成~~
~~1. 仮の経路情報を作る~~
~~2. ブロックチェーン上に刻む~~

~~そこからサーバーをFastAPIで構築~~

**Reply by tsmiyamoto** [reply]
やること
- 緯度経度と進行方向の情報から、こんなパノラマ画像を作成できますよ！という一つのデモを実際の道で作成
- 簡易のPythonスクリプトも作成

1つのデモを作るのは片方向ともう片方向からの写真で十分だと思うけど、発表用に動画で2方向から撮影して、この地点を切り取りました！的な見せ方ができるとよさそう
あまり人のいない道路などで撮影するとそれっぽくなるかもね

担当： @BB-Cotton

**Comment by tsmiyamoto** [comment]
**対向車と写真を合成して、あたかもストリートビューみたいな画像を作れる機能**
- ブロックチェーン上に刻まれているのは、マップ情報だけとする
- ブロックチェーン上から緯度経度及びその時の写真を撮ってくるプログラム
- 対向車との画像を繋ぎ合わせるアルゴリズム

**Comment by tsmiyamoto** [comment]
**フロント（ユーザー側にどう見せるか）**
- 自分の走行経路一覧
- 走行完了後に、コメントや写真を追加できる画面
- みんなのマップ（あんなところ行ってきた、こんなところ行ってきたが一覧になっているイメージ）
- 経路探索画面（みんなのマップを参照し、おすすめ経路をいくつか提案）

**Reply by tsmiyamoto** [reply]
@tsmiyamoto と @toryo-bobo の共同制作
Next.js頑張ろう！

**Reply by tsmiyamoto** [reply]
UI
https://www.bestnextjsthemes.com/theme/siumauricio-nextui-dashboard-template/

**Comment by tsmiyamoto** [comment]
基本はAWS上で動かしましょう
新しいアカウントを @tsmiyamoto が作る

**Comment by toryo-bobo** [comment]
### 経路情報について
- 緯度経度の情報を〇秒ごとにDB格納
- これをつなぎ合わせたものが経路情報になる

**Reply by toryo-bobo** [reply]
緯度経度の情報リストがブロックチェーン上に格納されて、経路としての価値を出す

**Reply by toryo-bobo** [reply]
## 緯度経度を取得してストレージ格納する方法について
- [Geolocation API](https://developer.mozilla.org/ja/docs/Web/API/Geolocation_API)というものがあるらしい。
  - ブラウザで位置情報の許可さえすれば、緯度経度を取得することができる

**Reply by toryo-bobo** [reply]
reference
[位置情報をブラウザで取得するGeolocation APIを使って今いる住所を調べる](https://webdrawer.net/javascript/geolocation.html)

**Comment by toryo-bobo** [comment]
### パノラマ画像について
- 車間で画像を共有してパノラマ画像を作成するのは難しい

**Reply by BB-Cotton** [reply]
> ・ 車間で画像を共有してパノラマ画像を作成するのは難しい

この理由は、写真の視点が一点でないと、違和感のないパノラマ画像を作るのが難しそう（今後の課題として、こういったパノラマ画像を作りたいはアリ）

**Reply by BB-Cotton** [reply]
・車両間で共有するのは、パノラマ画像を作るための画像ではなく、パノラマ画像自体

Googleストリートビューはパノラマ画像を無数に撮影することで実現している　→　本システムによって複数の車や監視カメラが撮影するパノラマ画像を共有することで、ストリートビューにする**パノラマ画像の数を補う**

**Reply by BB-Cotton** [reply]
・前後左右の四枚でパノラマ画像を作成しようとすると、3枚までしか反映されない（再現性：2回 / 2パターン中）
→ パノラマ画像にするには、画像をつなげる際の特徴量が十分にないとダメ。特徴量が少ない1枚があった？（要検証）

**Reply by BB-Cotton** [reply]
・Python+OpenCVのstitchingを利用してパノラマ画像を作成（参考：https://tech-blog.s-yoshiki.com/entry/36 ）

**Reply by tsmiyamoto** [reply]
### デプロイに関して
- AWS
- Serverless Framework
- FastAPI

あたりの技術スタックを使えば行けるはず

https://zenn.dev/hayata_yamamoto/articles/781efca1687272

**Reply by tsmiyamoto** [reply]
Input
~緯度・経度情報or~ timestamp
(ブロックチェーンと参照するため)

Output
パノラマ画像

**Reply by tsmiyamoto** [reply]
LambdaでOpenCVなどを使う方法
https://bftnagoya.hateblo.jp/entry/2021/06/23/074120

**Reply by tsmiyamoto** [reply]
LambdaでPNG画像を返す
https://qiita.com/edg_aim/items/e5c975fa35bbd12fd94a

**Reply by tsmiyamoto** [reply]
シナリオ整理
走行中の画像がtimestampと共に、S3に上がっている
ユーザーが「このときのパノラマ画像欲しい！」ってなったら、フロントからtimestampの情報をAPIでリクエストする
timestamp情報から、その時の緯度・経度と、S3に上がっている自身の画像および対向車の画像を取り出し、パノラマ画像を生成して返す

**Comment by toryo-bobo** [comment]
# Web3関係

ひとまず参考資料を載せておく（備忘録）
- [​1st Wave:Study Session](https://note.com/standenglish/n/nb52a762a695e)
- [動画](https://lu.ma/join/g-QILSqAXmxBbVSJA)
- [npm: @thirdweb-dev/react](https://www.npmjs.com/package/@thirdweb-dev/react)

**Reply by toryo-bobo** [reply]
## chainの変更方法

[thirdwebSDKのChain](https://portal.thirdweb.com/typescript/sdk.thirdwebsdk#chain-required)

```
import { ThirdwebSDK } from "@thirdweb-dev/sdk";
import { Gnosis } from "@thirdweb-dev/chains";

const sdk = new ThirdwebSDK(
  Gnosis,
);
```

**Reply by toryo-bobo** [reply]
## asterをMetaMaskに追加する方法
[ASTAR Network【ASTR】をメタマスクに追加する方法を紹介](https://www.sushitopmarketing.com/blog/astar-network-astr-metamask)

## そもそもMetaMaskってなんだよ
[MetaMask（メタマスク）とは](https://bitflyer.com/ja-jp/s/learn/metamask)

**Reply by toryo-bobo** [reply]
トークンバウンドアカウントの解説記事
[【完全保存版】ERC6551とは何か。トークンバウンドアカウントについて学びましょう。](https://zenn.dev/thirdweb_jp/articles/ce2c7bdc39ff85)

**Reply by toryo-bobo** [reply]
thirdwebを使うために
```
> yarn add @thirdweb-dev/react @thirdweb-dev/sdk ethers@5
```

**Comment by BB-Cotton** [comment]
**パノラマ画像を生成するLambda関数**

`s3://mapda3/<user>/<timestamp>`にパノラマ画像の元となる画像を保存する。
※デモ用に`<user>=user1, <timestamp>=demo`の中にtest1-1.JPEG, test1-2.JPEG, test1-3.JPEG, test1-4.JPEGという名前の画像を保存してある

jsonでtimestampの値をLambda関数に与えると、そのtimestampと同じフォルダ名の中の画像からパノラマ画像を生成する。
※デモ用に前述の`<user>`はuser1で固定。変更可。

生成したパノラマ画像は`s3://mapda3/<user>/<timestamp>`に保存。この画像の署名付きurlを戻り値とする。
※戻り値となるデータサイズが6MBを超えるとエラーとなるため、この方法に実装を変更。

**Reply by tsmiyamoto** [reply]
すばら 👏

**Reply by BB-Cotton** [reply]
この部分の実装終わったので、他に作った方がいいもの思い浮かんでたら教えて

**Reply by tsmiyamoto** [reply]
開発ではないんだけど、良ければThirdweb一通り使ってみてほしい
で、Python SDKもあるみたいだから、ワンチャンS3からIPFS（分散型ストレージ）に置き換えられたらベストなのかなと思ってる（が、まあここはオーバーな気もするので余裕があればぐらいで）

https://github.com/YAKIONIGIRIs/Mapda3/discussions/1#discussioncomment-7543388

**Reply by tsmiyamoto** [reply]
あと、Lambdaのエンドポイントと、API叩く参考プログラム（PythonなりcURLなり）欲しいかな

**Reply by tsmiyamoto** [reply]
@BB-Cotton 
Thirdwebも一通りと思ったけど、やっぱ下記をお願いしたい！
1. LambdaエンドポイントとAPI参考プログラム
2. 説明文（500文字以内）の草案作成
3. デモ動画の概要作成
  a. どういう構成にするか
  b. 何のアプリ使って動画作成するか（見せ方マジで大事なので、amazingな編集をしたい。が、自分にはそんな技術がないので、今どき流行りのAI動画製作アプリなんか使えないか。）
  
 提出物参考
 https://discord.com/channels/1062612424613691422/1143949101378043934/1175471787316891649

**Reply by tsmiyamoto** [reply]
木曜日にみんなで最終的な詰めをするが、草案作ってもらえるとだいぶ進みやすいかなと思って

**Reply by BB-Cotton** [reply]
Lamda関数のエンドポイント：https://r5gg3wiyb6.execute-api.ap-northeast-1.amazonaws.com/default/CreateImage
参考プログラム：/CreateMap/sample.py

**Reply by tsmiyamoto** [reply]
@BB-Cotton thank you
早起きね
